Added service.proto file and built thread pool functionality

GRaykovski opened the pull request.

radito3 commented on service.proto:
```diff
@@ -0,0 +1,55 @@
+syntax = "proto3";
+
+service Workers {
```
This isn't what our servce (and should be serviceS) should do. 
Take a look at the *.proto files in https://github.com/radito3/TicTacToe-android-client/tree/main/app/src/main/res

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
-    explicit ThreadPool(ThreadPool::Config config = {}) {} //TODO implement
+    explicit ThreadPool(ThreadPool::Config config = {}) {
+        start();
+    } //TODO implement
```
Why is this comment still present? Isn't the constructor implemented?

radito3 replied:
Do not resolve comments that haven't been fixed.

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
+        
         //FIXME temporary impl
-        threads.emplace_back(std::forward<std::function<void()>>(job));
+        //threads.emplace_back(std::forward<std::function<void()>>(job));
```
Remove code that won't be used

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
+        //threads.emplace_back(std::forward<std::function<void()>>(job));
         //if accepted
-        return true;
+        if (task_queue::size() < Config::task_queue_size) {
```
operator:: isn't the proper way to use an instance method. It's used to access static class fields / methods.

radito3 replied:
Do not use the Config object's properties. That object's lifetime ends with the construction of the ThreadPool.
Generally, configuration objects are used to set class fields, as an alternative to a builder, not to be used by those objects after creation.

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
-        return true;
+        if (task_queue::size() < Config::task_queue_size) {
+
+            auto wrapper = std::make_shared<std::packaged_task<decltype(job()) ()>>(std::move(job()));
```
- Why do you need to use `decltype`? The return type of the functor is known -> void
- This isn't asynchronous execution. You are calling the `job` object and moving its result. This will be either a compile time error (since void doesn't define operator()) or a runtime error because `packaged_task` expects a callable object (like `function`).
- Why do you need a pointer to the packaged task?

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
+            {
+                std::unique_lock<std::mutex> lock{ event_mutex };
+                task_queue.emplace([=] {
+                    (*wrapper)();
```
This lambda won't do much, since the job has already been called and has blocked the current thread.

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
+
+            {
+                std::unique_lock<std::mutex> lock{ event_mutex };
+                task_queue.emplace([=] {
```
Why are you capturing by value?

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
     }
 
-    bool submit_job(std::function<void()>&& job) {
+    auto submit_job(std::function<void()>&& job)->std::future<decltype(job())> 
```
The return type is known beforehand. `std::future<void>` You don't need to use this syntax

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
         //if rejected
-        return false;
+        else {
+
```
Why is this block empty?

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
+
+                    std::function<void()> job;
+
+                    {
```
This is too much indentation. Refactor.

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
+
+    void start(num_active_threads) {
+        for (auto i = 0u ; i < num_active_threads; ++i) {
+            threads.emplace_back([=] {
```
Again, why are you capturing by value?

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
+  
+
+    void start(num_active_threads) {
+        for (auto i = 0u ; i < num_active_threads; ++i) {
```
Why do you need an integer suffix and `auto` for a simple for loop?

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
+                        std::unique_lock<std::mutex> lock{ event_mutex };
+                        event_var.wait(lock, [=] {return stoping || !task_queue.empty(); });
+
+                        if (stoping && task_queue.empty())
```
What if the stop flag is set but the queue isn't empty? Memory leak?

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
+
+                    }
+
+                    job();
```
This isn't asynchronous execution.

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
 
+  
+
+    void start(num_active_threads) {
```
After having a look through this method's definition, this is a blocking method.
And since it's called in the object's constructor, this object will block its calling thread upon construction

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
+            })
+        }
+    }
+    void stop() noexcept
```
Why is this method marked `noexcept`? Can't joining the threads throw an exception?

radito3 commented on thread_pool.h:
```diff
@@ -25,35 +30,97 @@ class ThreadPool {
     //FIXME temporary impl
     std::vector<std::thread> threads;
+    bool stoping = false;
+    std::condition_variable event_var;
```
This variable name doesn't mean much. Rename it.

radito3 reviewed (changes requested):
The gRPC usage is incorrect.
Even though the idea of a thread pool may have been grasped, the implementation isn't correct.

GRaykovski committed:
Add protobuf definitions & work on impl

- Work on thread pool
- Fix HttpServer::register_handler
- Create search, lobby & game sessions managers
- Add grpc to cmake build
- Update Dockerfile
- Add gRPC clients
- Add gRPC delegating implementations of I/O interface

radito3 commented on Coordinate.proto:
```diff
@@ -0,0 +1,6 @@
+message Coordinate {
+     int32 x = 1;
+     int32 y = 2;
+}
```
Add a new line at the end of files.

radito3 commented on GameSessionsManager.h:
```diff
@@ -0,0 +1,22 @@
+#include <GameSession.h>
+
+class GameSessionsManager {
+	ThreadPool threadpool;
```
This doesn't follow naming conventions. Change it.

radito3 commented on GameSessionsManager.h:
```diff
@@ -0,0 +1,22 @@
+
+	void start_game_session(SearchManager::PlayerInfo player1, SearchManager::PlayerInfo player2) {
+		threadpool.submit_job([&]() {
+			GameSession session(Player(player1.id, symbol::CROSS, new GrpcDelegatingWriter(player1.adress), new GrpcDelegatingReader(player1.adress)),
```
Extract player creation before GameSession and pass by constant reference. This line is too cluttered.

radito3 commented on LobbyManager.h:
```diff
@@ -0,0 +1,60 @@
+	struct Lobby {
+		SearchManager::PlayerInfo player1;
+		SearchManager::PlayerInfo player2;
+		bool player1_active;
+		bool player2_active;
```
These two fields shouldn't be visible outside this class. Refactor it so that they aren't returned from a public method

radito3 replied:
This isn't resolved. Fix it.

radito3 commented on LobbyManager.h:
```diff
@@ -0,0 +1,60 @@
+	};
+
+private:
+	unordered_map<unsigned, Lobby> active_lobbies;
```
This is a compile-time error. There is no class/struct/union named "unordered_map" in the current namespace.

radito3 commented on LobbyManager.h:
```diff
@@ -0,0 +1,60 @@
+public:
+
+	void create_lobby(SearchManager::PlayerInfo player1, SearchManager::PlayerInfo player2, unsigned id) {
+		active_lobby[std::mt19937(std::random_device()())()] = Lobby(player1, player2, id);
```
This is a logical error. Why are you passing an id (and `id` is too generic. Rename it.) but then generating a random key for the map entry?

radito3 commented on LobbyManager.h:
```diff
@@ -0,0 +1,60 @@
+		active_lobby[std::mt19937(std::random_device()())()] = Lobby(player1, player2, id);
+	}
+
+	bool check_for_active_lobby_with_id(std::string player_id ) {
```
The name of this method doesn't match it's functionality & input parameters. The name suggests that we pass the lobby id, whereas we are passing a player id. Rename or refactor.

radito3 commented on SearchManager.h:
```diff
@@ -0,0 +1,37 @@
+
+class SearchManager {
+public:
+	struct PlayerInfo {
```
Extract this to its own file

radito3 commented on SearchManager.h:
```diff
@@ -0,0 +1,37 @@
+public:
+	struct PlayerInfo {
+		std::string id;
+		std::string adress;
```
address*

radito3 commented on SearchManager.h:
```diff
@@ -0,0 +1,37 @@
+	};
+
+private:
+	unordered_map<unsigned, PlayerInfo> active_searches;
```
Same error as before. No class/struct/union with this name in the current namespace

radito3 commented on SearchManager.h:
```diff
@@ -0,0 +1,37 @@
+		}
+	}
+
+	PlayerInfo get_player_info() {
```
- This is a compile-time error. A map's iterator's value is a pair (key + value), not just the value
- Why do you only need the player info without the search id?

radito3 replied:
This isn't resolved.

radito3 commented on grpc/DisplayWorkerClient.h:
```diff
@@ -0,0 +1,139 @@
+
+class DisplayWorkerClient {
+	std::unique_ptr<DisplayWriter::Stub> stub;
+	google::protobuf::Empty empty;
```
This name doesn't mean anything. Rename to something like `empty_request`

radito3 commented on grpc/DisplayWorkerClient.h:
```diff
@@ -0,0 +1,139 @@
+
+	void write_grid() const {
+		ClientContext context;
+		google::protobuf::Empty e;
```
This name doesn't mean anything. Rename to something like `empty_response`

radito3 commented on grpc/DisplayWorkerClient.h:
```diff
@@ -0,0 +1,139 @@
+
+	void clear_cell_at(const coordinate& coordinate) const {
+		ClientContext context;
+		google::protobuf::Empty e;
```
Same for here and all of the following in this class

radito3 commented on grpc/InputReaderClient.h:
```diff
@@ -0,0 +1,44 @@
+#ifndef TICTACTOE_SERVER_INPUTREADERCLIENT_H
```
Fix the indentation in this file

radito3 commented on grpc/InputReaderClient.h:
```diff
@@ -0,0 +1,44 @@
+			google::protobuf::Empty e;
+			Status status = stub->read(&context, e, &input);
+
+			if (status.ok()) {
```
Invert this conditional to reduce indentaion

radito3 commented on thread_pool.h:
```diff
@@ -54,18 +66,75 @@ class ThreadPool {
 //        rejected_job_policy = policy;
 //    }
 
+>>>>>>> 14dde82 (Add protobuf definitions & work on impl)
```
Remove this line

GRaykovski committed:
Fix bugs and refactor code

More refactoring

radito3 commented on GameSessionsManager.h:
```diff
@@ -1,19 +1,25 @@
 
 class GameSessionsManager {
-	ThreadPool threadpool;
+	ThreadPool game_session_threads;
```
This name implies that there is only 1 game session with multiple threads. Rename.

radito3 commented on GameSessionsManager.h:
```diff
@@ -1,19 +1,25 @@
+	ThreadPool game_session_threads;
 
-public:
+	Player create_player(const std::string& player_id, const symbol& symbol_, const std::string& player_address){
```
What is this factory method for? And `symbol` is an enum. Passing it by value or const reference doesn't make much a difference in terms of memory

radito3 commented on GameSessionsManager.h:
```diff
@@ -1,19 +1,25 @@
+public:
+	void start_game_session(PlayerInfo player1, PlayerInfo player2) {
+		game_session_threads.submit_job([&]() {
+			GameSession session(create_player(player1.id, symbol::CROSS, player1.address),
```
This isn't what I meant with my previous comment about this line. Refactor it to enhance readability. This line contains too much information.

radito3 commented on LobbyManager.h:
```diff
@@ -1,44 +1,45 @@
 		unsigned search_id;
 
-		Lobby(SearchManager::PlayerInfo player1, SearchManager::PlayerInfo player2, unsigned id) : player1(player1), player2(player2), search_id(id){}
+		Lobby(PlayerInfo player1, PlayerInfo player2, unsigned lobby_id) : player1(player1), player2(player2), search_id(lobby_id){}
```
This is a naming mismatch. Why are you passing `lobby_id` to a field named `search_id`?

radito3 commented on LobbyManager.h:
```diff
@@ -1,44 +1,45 @@
 			return false;
-		} else if (lobby.player2.id == player_id){
-			lobby.player2_active = true;
+		} else if (lobby.second.player2.id == player_id){
```
More than 1 `else if` statement becomes hard to read. Refactor.

radito3 commented on LobbyManager.h:
```diff
@@ -48,11 +49,12 @@ class LobbyManager {
 		} 
-		return active_lobbies.at(id);
+		auto requested_lobby = active_lobbies.at(id);
+		return {requested_lobby.player1, requested_lobby.player2, requested_lobby.search_id};
```
Again, this isn't what I meant with my previous comment. The fields specifying whether the players are active are still publicly accessible outside this class.

radito3 replied:
Again, you haven't changed the root of the issue. In fact, you've made it worse. You have moved information designated to the lobby to the `PlayerInfo` struct which doesn't need it.

radito3 commented on PlayerInfo.h:
```diff
@@ -0,0 +1,11 @@
+#include <string>
+
+struct PlayerInfo {
+		std::string id;
+		std::string address;
```
What is this indendation?

radito3 commented on PlayerInfo.h:
```diff
@@ -0,0 +1,11 @@
+		std::string address;
+};
+
+#endif
```
No new line at end of file...

radito3 commented on SearchManager.h:
```diff
@@ -3,16 +3,11 @@
 
 #include <unordered_map>
 #include <random>
+#include <PlayerInfo.h>
```
Local files shouldn't be imported with `<>`.

radito3 commented on SearchManager.h:
```diff
@@ -25,8 +20,9 @@ class SearchManager {
 
-	PlayerInfo get_player_info() {
-		return *active_searches.begin();
+	PlayerInfo get_player() {
```
This name implies that the class has a field of `Player` or some subclass of Player. Change it.

radito3 commented on grpc/InputReaderClient.h:
```diff
@@ -10,35 +10,42 @@ using grpc::Channel;
 
-class InputReaderClient {
+class InputReaderClient
+{
```
This doesn't follow the formatting style of this project.

radito3 commented on grpc/InputReaderClient.h:
```diff
@@ -10,35 +10,42 @@ using grpc::Channel;
-			google::protobuf::Empty e;
-			Status status = stub->read(&context, e, &input);
+	input_t read() const
+	{
```
Neither does this.

radito3 commented on grpc/InputReaderClient.h:
```diff
@@ -10,35 +10,42 @@ using grpc::Channel;
-				std::cerr << status.error_code() << ": " << status.error_message() << std::endl;
-				return input_t(move_direction::INVALID);
+		if (!status.ok())
+		{
```
.

radito3 commented on grpc/InputReaderClient.h:
```diff
@@ -10,35 +10,42 @@ using grpc::Channel;
+			std::cerr << status.error_code() << ": " << status.error_message() << std::endl;
+			return input_t(move_direction::INVALID);
+		}
+		else
```
This is C#-esque style of formatting.

radito3 commented on grpc/InputReaderClient.h:
```diff
@@ -10,35 +10,42 @@ using grpc::Channel;
 			}
+			return input_t(move_direction::INVALID);
 		}
+	}
```
This isn't what I meant by my previous comment. The indentation of this block of code hasn't changed. Inverting the if statement has no effect with this change.

radito3 commented on CMakeLists.txt:
```diff
@@ -1,12 +1,21 @@
 
-add_executable(TicTacToe_server main.cpp HttpServer.cpp HttpServer.h request_matcher.h http_methods.h http_response.h http_request.h http_request_context.h thread_pool.h io/GrpcDelegatingWriter.h io/GrpcDelegatingReader.h)
-target_link_libraries(${PROJECT_NAME} TicTacToe)
+add_executable(TicTacToe_server main.cpp HttpServer.cpp HttpServer.h request_matcher.h http_methods.h http_response.h http_request.h http_request_context.h thread_pool.h io/GrpcDelegatingWriter.h io/GrpcDelegatingReader.h DisplayWriterService.grpc.pb.cc DisplayWriterService.grpc.pb.h DisplayWriterService.pb.cc DisplayWriterService.pb.h InputReaderService.grpc.pb.h InputReaderService.grpc.pb.cc In ...
```
remove the quotes around the new file names

radito3 commented on DisplayWriterService.proto:
```diff
@@ -0,0 +1,49 @@
+
+message TextMessage {
+     string value = 1;
+}
```
new line

radito3 commented on GameSessionsManager.h:
```diff
@@ -0,0 +1,26 @@
+	}
+};
+
+#endif
```
new line

radito3 commented on SearchManager.h:
```diff
@@ -0,0 +1,33 @@
+		if (active_searches.empty()) {
+			active_searches[std::mt19937(std::random_device()())()] = player ;
+			return false;
+		} else {
```
No need for this `else` statement. Just return after the `if` body.

radito3 commented on grpc/DisplayWorkerClient.h:
```diff
@@ -0,0 +1,139 @@
+		Status status = stub->write_temp_msg(&context, text_msg, &empty_response);
+
+		if (!status.ok()) {
+			std::cerr << status.error_code() << ": " << status.error_message() << std::endl;
```
This line is repeated for every method. Overload the operator `<<` for `Status`

radito3 commented on thread_pool.h:
```diff
@@ -55,17 +67,72 @@ class ThreadPool {
 };
 
-#endif //TICTACTOE_SERVER_THREAD_POOL_H
+#endif //TICTACTOE_SERVER_THREAD_POOL_H
```
why remove the trailing new line?

radito3 commented on GameSessionsManager.h:
```diff
@@ -0,0 +1,26 @@
+		const Player& first_player = Player(player1.id, symbol::CROSS, new GrpcDelegatingWriter(player1.address), new GrpcDelegatingReader(player1.address));
+		const Player& second_player = Player(player2.id, symbol::CIRCLE, new GrpcDelegatingWriter(player2.address), new GrpcDelegatingReader(player2.address));
+
+		game_sessions_pool.submit_job([&]() {
+			GameSession session(first_player, second_player);
+			session.play();
+		});
```
The lambda expression will outlive the variables it has been passed, so capturing by reference will produce dangling references.

radito3 commented on GameSessionsManager.h:
```diff
@@ -0,0 +1,26 @@
+
+public:
+	void start_game_session(PlayerInfo player1, PlayerInfo player2) {
+		const Player& first_player = Player(player1.id, symbol::CROSS, new GrpcDelegatingWriter(player1.address), new GrpcDelegatingReader(player1.address));
+		const Player& second_player = Player(player2.id, symbol::CIRCLE, new GrpcDelegatingWriter(player2.address), new GrpcDelegatingReader(player2.address));
```
call the constructors directly. there is no need to assign an rvalue to a const reference.

radito3 commented on PlayerInfo.h:
```diff
@@ -0,0 +1,12 @@
+struct PlayerInfo {
+	std::string id;
+	std::string address;
+	bool is_active;
```
Should this information be present here? What does it mean for a player info object to be "active"?

radito3 commented on SearchManager.h:
```diff
@@ -0,0 +1,32 @@
+
+	std::pair<unsigned, PlayerInfo> get_player_info_with_search_id() {
+		auto player = *active_searches.begin();
+		return player;
```
This is a compilation error...
